plugins/modules/foreman_compute_resource.py

```diff
@@ -112,133 +154,93 @@
-
-# This is the only true source for names (and conversions thereof)
-name_map = {
-    'name': 'name',
-    'description': 'description',
-    'provider': 'provider',
-    'organizations': 'organization',
-    'locations': 'location',
-    'datacenter': 'datacenter',
-}
+from ansible.module_utils.foreman_helper import ForemanEntityApypieAnsibleModule
 
 
 def get_provider_infos(provider):
     provider_name = provider.lower()
 
     if provider_name == 'libvirt':
         return {
-            'params': ['url', 'display_type'],
-            'class': LibvirtComputeResource
+            'params': ['url', 'display_type']
         }
 
     elif provider_name == 'ovirt':
         return {
-            'params': ['url', 'user', 'password', 'datacenter', 'use_v4', 'ovirt_quota'],
-            'class': OVirtComputeResource
+            'params': ['url', 'user', 'password', 'datacenter', 'use_v4', 'ovirt_quota']
         }
 
     elif provider_name == 'vmware':
         return {
-            'params': ['url', 'user', 'password', 'datacenter'],
-            'class': VMWareComputeResource
+            'params': ['url', 'user', 'password', 'datacenter']
         }
 
     else:
         return {
-            'params': [],
-            'class': AbstractComputeResource
+            'params': []
         }
 
 
 def main():
-    module = ForemanEntityAnsibleModule(
-        argument_spec=dict(
-            name=dict(type='str', required=True),
-            updated_name=dict(type='str'),
-            description=dict(type='str'),
-            provider=dict(type='str', choices=['vmware', 'libvirt', 'ovirt']),
+    module = ForemanEntityApypieAnsibleModule(
+        entity_spec=dict(
+            name=dict(required=True),
+            updated_name=dict(),
+            description=dict(),
+            organizations=dict(type='entity_list', flat_name='organization_ids'),
+            locations=dict(type='entity_list', flat_name='location_ids'),
+            provider=dict(choices=['vmware', 'libvirt', 'ovirt']),
             provider_params=dict(type='dict'),
-            locations=dict(type='list'),
-            organizations=dict(type='list'),
-            state=dict(type='str', default='present', choices=['present', 'absent', 'present_with_defaults']),
-
-            # Deprecated provider's specific params, use nested keys in provider_params param instead
-            provider_auth=dict(type='dict'),
-            url=dict(type='str'),
-            display_type=dict(type='str'),
-            datacenter=dict(type='str'),
+            display_type=dict(),
+            datacenter=dict(),
+            url=dict(),
+            user=dict(),
+            password=dict(),
+            use_v4=dict(),
+            ovirt_quota=dict(),
+        ),
+        argument_spec=dict(
+            parameters=dict(type='list', elements='dict', options=dict(
+                state=dict(type='str', default='present', choices=['present', 'absent', 'present_with_defaults']),
```

Comment:
The state should really be a direct member of the `argument_spec`. Also i do not see parameters anywhere in the Documentation.